org.eclipse.lsp4e/src/org/eclipse/lsp4e/LanguageServerWrapper.java

```diff
@@ -224,6 +224,26 @@ public synchronized void start() throws IOException {
 			}
 		}
 		if (this.initializeFuture == null ) {
+			final URI rootURI;
+			final IProject project = this.initialProject;
+			if (project != null && project.exists()) {
+				rootURI = LSPEclipseUtils.toUri(this.initialProject);
+			} else {
+				// Introducing this local variable is required due to overzealous static analysis.
+				// Dereferencing this.initialPath directly will trigger a "potential null" warning/error.
+				// Checking for this.initialPath == null is not enough.
+				final IPath initialPath = this.initialPath;
+				if (initialPath != null) {
+					File projectDirectory = initialPath.toFile();
+					if (projectDirectory.isFile()) {
+						projectDirectory = projectDirectory.getParentFile();
+					}
+					rootURI = LSPEclipseUtils.toUri(projectDirectory);
+				} else {
+					rootURI = LSPEclipseUtils.toUri(new File("/")); //$NON-NLS-1$
+				}
+			}
```

Comment:
I understand that you are just moving the code, but if you would move this block to a method such as
```
private URI getRootUri(IPath initialPath) {
  ...
}
```
you could remove the three lines comment and the code would be cleaner.

Reply:
+1 this would be welcome.

Reply:
The comment is only about the first line of the block. keeping or removing it has nothing to do with putting that block in a method. For someone who worked eclipse null analysis the comment is superficial. So for me this is as meaningful as a comment above a variable assignment that describes that in the next line a value is assigned to a the variable on the left side. So I am all in for removing the comment but I don't really see value in putting the block into a separate method.

Reply:
In the end it is only about taste, to me the method call looks like a more natural way to remove the null analysis complaint. Probably nobody would wonder why there is a method call, while now apparently someone wondered (or thought others would wonder and added the comment) why there is an assign such as `final IPath initialPath = this.initialPath;`.

In any case, it is not very important, so I would suggest we forget it ;)

Still, while looking at the code more carefully I have realize the same type of assignment is done on line 250 `IProject project = this.initialProject;`, what about moving the comment before that line so that it appears before its first occurrence and not the second?

Also, line 249 `URI rootURI = null` can be moved inside the if block, and then, the block should use project, not this.initialProjoct. All together we would have

```
					// This is required due to overzealous static analysis. Dereferencing
					// this.initialProject / this.initialPath directly will trigger a "potential null"
					// warning/error. Checking for this.<member> == null is not enough.					
					IProject project = this.initialProject;
					if (project != null && project.exists()) {
						URI rootURI = LSPEclipseUtils.toUri(project);
						initParams.setRootUri(rootURI.toString());
						initParams.setRootPath(rootURI.getPath());
					} else {

```

Would you mind doing that as part of the PR?

BTW: The change makes sense to me.

Reply:
How about like this? https://github.com/eclipse/lsp4e/pull/158/commits/c44abe29c93bdb2cab02343ef5728b87d6d4a6bd